Content.Client/RCD/RCDConstructionGhostSystem.cs

```diff
@@ -0,0 +1,253 @@
+    private ProtoId<RCDPrototype>? _constructionPrototype = null;
+    private RCDPrototype? _cachedPrototype = null;
+    private bool _isKeyBindActive = false;
+
+    private readonly BoundKeyFunction _keyfunction = EngineKeyFunctions.EditorRotateObject;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        // Determine if player is carrying an RCD in their active hand
+        var player = _playerManager.LocalSession?.AttachedEntity;
+
+        if (!TryComp<HandsComponent>(player, out var hands))
+            return;
+
+        var uid = hands.ActiveHand?.HeldEntity;
+        var hasRCD = TryComp<RCDComponent>(uid, out var rcd);
+
+        // Rotate the construction ghost if the keybind was actived this frame
+        if (WasKeybindActivatedThisFrame(player.Value) && hasRCD)
+            RotateRCDConstructionGhost(uid!.Value, rcd!);
+
+        // Delete the construction ghost if its no longer needed
+        if (!hasRCD ||
+            !EntityManager.TryGetComponent<InputComponent>(player, out var input) ||
+            _inputManager.Contexts.ActiveContext.Name != input.ContextName)
+        {
+            DeleteConstructionGhost();
+            return;
+        }
+
+        if (_constructionPrototype != rcd!.ProtoId)
+        {
+            DeleteConstructionGhost();
+
+            _constructionPrototype = rcd!.ProtoId;
+            _protoManager.TryIndex(rcd!.ProtoId, out _cachedPrototype);
+        }
+
+        if (_cachedPrototype?.GhostSprite == null)
+        {
+            DeleteConstructionGhost();
+            return;
+        }
+
+        // Try to get the tile the player is hovering over
+        if (!CurrentMousePosition(player.Value, out var screenCoords))
+            return;
+
+        var location = ScreenToCursorGrid(screenCoords);
+
+        if (!_rcdSystem.TryGetMapGridData(location, out var mapGridData))
+            return;
+
+        // Delete the construction ghost if the distance between the player and the target tile is too great
+        var tileWorldPosition = _mapSystem.GridTileToWorldPos(mapGridData.Value.GridUid, mapGridData.Value.Component, mapGridData.Value.Position);
+        var distance = (tileWorldPosition - _transformSystem.GetWorldPosition(player.Value)).Length();
+
+        if (distance > 1.5f)
+        {
+            DeleteConstructionGhost();
+            return;
+        }
+
+        // Create a new construction ghost if needed
+        SpriteComponent? sprite;
+        var tilePosition = _mapSystem.GridTileToLocal(mapGridData.Value.GridUid, mapGridData.Value.Component, mapGridData.Value.Position);
+
+        if (_constructionGhost == null)
+        {
+            _constructionGhost = EntityManager.SpawnEntity("rcdconstructionghost", tilePosition);
+
+            sprite = EntityManager.GetComponent<SpriteComponent>(_constructionGhost.Value);
```

Comment:
just do a trycomp since you really shouldn't assume that everything has a sprite component.